src/chart/helper/createSeriesDataFromArray.ts

```diff
@@ -107,26 +99,92 @@ function createListFromArray(source: Source | OptionSourceData, seriesModel: Ser
+    }
+): SeriesData {
+    opt = opt || {};
 
-    const list = new List(dimInfoList, seriesModel);
+    if (!isSourceInstance(sourceOrStore) && !isDataStorage(sourceOrStore)) {
+        sourceOrStore = createSourceFromSeriesDataOption(
+            sourceOrStore as OptionSourceData
+        );
+    }
 
-    list.setCalculationInfo(stackCalculationInfo);
+    const source = isDataStorage(sourceOrStore) ? sourceOrStore.getSource() : sourceOrStore;
+    const coordSysInfo = getCoordSysInfoBySeries(seriesModel);
+    const coordSysDimDefs = getCoordSysDimDefs(seriesModel, coordSysInfo);
+    const useEncodeDefaulter = opt.useEncodeDefaulter;
 
-    const dimValueGetter = (firstCategoryDimIndex != null && isNeedCompleteOrdinalData(source))
-        ? function (this: List, itemOpt: any, dimName: string, dataIndex: number, dimIndex: number) {
-            // Use dataIndex as ordinal value in categoryAxis
-            return dimIndex === firstCategoryDimIndex
-                ? dataIndex
-                : this.defaultDimValueGetter(itemOpt, dimName, dataIndex, dimIndex);
-        }
+    // Try to ignore unsed dimensions if sharing a high dimension datastorage
+    // 30 is an experience value.
+    const omitUnusedDimensions = isDataStorage(sourceOrStore) && sourceOrStore.getDimensionCount() > 30;
+    const encodeDefaulter = zrUtil.isFunction(useEncodeDefaulter)
+        ? useEncodeDefaulter
+        : useEncodeDefaulter
+        ? zrUtil.curry(makeSeriesEncodeForAxisCoordSys, coordSysDimDefs, seriesModel)
         : null;
 
-    list.hasItemOption = false;
-    list.initData(source, null, dimValueGetter);
+    const createDimensionOptions = {
+        coordDimensions: coordSysDimDefs,
+        generateCoord: opt.generateCoord,
+        encodeDefine: seriesModel.getEncode()
+            // NOTE: If we call createDimensions on same source multiple times.
+            // It will break the encodeDefaulter which has sideeffects.
+            // So we prepare the default encode here instead of passing encoderDefaulter function.
+            || (encodeDefaulter && encodeDefaulter(
+                source, getDimCount(source, coordSysDimDefs, source.dimensionsDefine || [])
+            )),
+        omitUnusedDimensions
+    };
+    let dimInfoList = createDimensions(sourceOrStore, createDimensionOptions);
+    let firstCategoryDimIndex = injectOrdinalMeta(dimInfoList, opt.createInvertedIndices, coordSysInfo);
+
+    if (omitUnusedDimensions) {
+        // sourceOrStore
+        if (!(sourceOrStore as DataStorage).syncDimensionTypes(dimInfoList)) {
+            dimInfoList = createDimensions(sourceOrStore, zrUtil.extend(createDimensionOptions, {
+                omitUnusedDimensions: true
+            }));
+            // Fallback
+            firstCategoryDimIndex = injectOrdinalMeta(
+                dimInfoList, opt.createInvertedIndices, coordSysInfo
+            );
+            sourceOrStore = source;
+        }
+    }
```

Comment:
This code will try to create a dimensions list that only include needed and use `syncDimensionTypes(dimInfoList)` to sync to the storage. If fails to sync, it will recreate a full dimensions list which will be used when recreating the DataStorage. I updated the comment a bit to explain it.